Tablenamer

derkan opened the pull request:
* Define table name for structs. (You can call one of `SetTableNamer...` functions at your main function to use a namer.) For example, if struct's name is `BookAuthor`:
    * Default tablename will be same as struct's name(`SetTableNamerSame()`). Example table name will be `BookAuthor`.
    * Call `SetTableNamerPlural()` to use plural name of struct. Example table name will be `BookAuthors`.
    * Call `SetTableNamerSnake()` to get struct name as sname format. Example table name will be `book_author`.
    * Call `SetTableNamerSnakePlural()` to get struct name as plural snake format. Example  table name will be `book_authors`.
    * If none of options is good for you add `TableName() string` method to for your struct and return whatever table name will be.

samonzeweb commented:
Hi Derkan,

I must admit that this PR poses me some problems.

Technically the main problem is the use of global settings. Using a single DB is a common case, but not the only one. Different DB could have different naming rules. You can't simply fallback to `TableName()`, because this implies that a data layer package will behave differently according to the context. Then the only way to be safe is to always use `TableName()`.

Of course this problem disappear if the settings is at the connection level (a `godb.DB` struct).

Another problem is the constant computation of table names. It's something which could be cached (I don't know where now, just a thought). But it's also something wich must not simply be stored un any `godb.DB` struct. A web server will clone the struct for each incoming request, and the cache will start empty (or almost empty) all the time. I must think about safe shared state across different instances sharing the same *root* struct. And also add security to avoid logical cache corruption in case of call to a SetTableNamerXXXX function after the first entry was added to the cache.

Other problem (?) is the use of `github.com/jinzhu/inflection` package. There is no legal (licence) problem, but I don't know if it's fair play as it was developed to be used in `gorm` . Maybe I'm asking myself too many questions. Perheps it's not a problem at all.

About `godb` and table names : initially I made a choice of **not** having any features like this, and adopt the Python mantra : *explicit is better than implicit*. Then there is no default field/column mapping, the column name associated to each field is mandatory. For the same reason I only made a simple default case for table name : the struct name (as is). Other kind of table name has to be explicitly specified with `TableName()`.

Let me some time to think about all these points and read with more attention the code (I'm still very busy : work and personal reasons). But at least I made a first reply ;)

Sam.

derkan commented:
Hi Sam,
Thanks for review. Here my opinions:
- Global setting problem: No problem to carry global var to `godb.DB`
- Constant computation of table names problem:  TableNamerFunc can be restricted to be called once. And raise error if it is called second time. Or can be set only during connection to force it not be changed later. We can call this prop to DefaultTableNamer function, as it will be set once.
- jinzhu's inflection package: I've used that because of it's MIT licence, and also I don't like dependecies and looking an easy way to develop function only for pluralizing. :-)
- For each struct, adding an extra method for naming tables seems overwhelmed to me. For struct naming singular version is suggested. From DBA's perspective table names should be plural. So, default 
`godb` naming falls short here. Because of this I added table naming options.
Erkan.

derkan commented:
Hi Sam,
I refactored TableNamer PR. Changes are:
- `tablenamer` is a subpackage now.
- Removed dependency for `github.com/jinzhu/inflection` added custom and simpler `plural` package
- `godb.defaultTableNamer` is used to hold active `TableNamerFn`.
- `godb.SetDefaultTableNamer(tablenamer.TableNamerFn)` function should be called to set a tablenamer from `tablenamer` subpackage or a custom namer function.
Erkan

derkan commented:
Hi @samonzeweb ,
Did you have a chance to review after latest changes?

samonzeweb commented:
Hi,

I'm sorry' I missed the notification 13 days ago and was on holiday last week (less connected). I really apologize.

I'll try to do a review before next week.

samonzeweb commented:
# First/fast review

I made a first review. Not about all plural/snake/... details, but rather the overall design. I didn't read tests details.

I initialy write some remarks I removed. Questions about public function in plural package could change many things. I thing this point should be clear before doing any refactoring.

I thing I have to take time to re-read things after your reply. But as things have already took too long time I prefer give you a first/'rapid' feedback.


## README

Too many details. It has to be concise, it's better to create a dedicated chapter in documentation.


## tablenamer

### getTableName()

Add comment about the returned boolean. Even if it's *private* code, it's not obvious here for future maintainer (I only understand the purpose reading other parts).


### Same pattern multiple times 

Perhaps something to refactor ?

```
db.defaultTableNamer(si.recordDescription.getTableName())
```

A (probably bad) idea is to delegate the table naming to the recordDescription as it's the case now, but it requires either a pointer to godb or the defaultTableNamer. Drawback : more coopling. 

Not a big problem, just a thing which could probably be improved. Without a really good idea it's better to let it as-is for now.


### ToSnakeCase()

Why it's public ?


### Caching / memoization

All methods having a significant cost should cache already computed values. It's useless and costly to computed many times the plural of the same word, or convert it to snake case.


## Plural

### Public functions

There are many publics functions. Is it a choice ? How to use them in real case ?

This question is **very** important because some design part could change according to this question (and then I could add more remarks).


### Many iterations

You use many `range` where perhaps you could use a prepopulated map (even a map[string]bool if you just check if the value exists). 

### Wrong test name

In `enplural_test.go` : `func TestJSONStr(t *testing.T) {`

derkan commented:
Hi Sam,

All good points, thanks for reviewing. I've refactored as:
- tablenamer:
   - Same pattern multiple times: Yes, I couldn't find a way to refactor it, as too much refactoring may be needed on `godb/struct*` files.
   -  `getTableName()`: added comment for bool
   - `ToSnakeCase()`: Public because user may define his custom method for naming tables and can call this.
   - Caching / memoization: snake casing is cached, as plural is cached in plural package, it doesn't need caching.
- Wrong test name:
   - Corrected
- Plural:
  - Public Functions:
     - `Add*` may be needed for enduser to add custom conditions, so left public on purpose. Only `GetIf*` and `isUncountable` are now private. `Plural` is public as it can be used in user defined `tablenamer`s.
   - Many iterations:
      -  `getIfPlural`:  input value should be checked against  all regex until match, so range is needed. But `uncountables` and `irregulars` are now map.
      - Added cache.

samonzeweb commented on tablenamer/tablenamer.go:
```diff
@@ -0,0 +1,115 @@
+type TableNamerFn func(string, bool) string
+
+// snakeCache is cacher for ToSnakeCase function
+type snakeCache struct {
```
This cache will have very few writes and many reads. For avoid performance issues due to lock it's better to use https://golang.org/pkg/sync/#Map
It will probably break compatibility with Go 1.8 , but its support will end in two months.

samonzeweb commented on tablenamer/tablenamer.go:
```diff
@@ -0,0 +1,115 @@
+	"sync"
+)
+
+// TableNamer function type is used set table naming function to build table name from struct's name
```
Bad type name in comment (thanks vscode and its excellent Go tools integrationt).

samonzeweb commented on tablenamer/plural/enplural.go:
```diff
@@ -0,0 +1,198 @@
+}
+
+// EnglishPlural defines methods and rules for getting plural form of a word for English
+type EnglishPlural struct {
```
The cache will be more efficient with https://golang.org/pkg/sync/#Map
(see another similar comment in this review)

I wonder about the need to lock things while adding / accessing rules, I think it's better to do it one **before** computing any table name. I'll explain it in a comment into the PR.

samonzeweb commented:
Hi Derkan,

I made some comments in a code review.

Concerning ` AddPluralRegex` and its triends, how can somebody use it to add custom rules ? Each time you set a default table namer you create a specific instance, which is not accessible. I also made a remarks about the lock strategy to manage plural rules.

* If you have a package level variable containing only one instance of plural rules you can more easily add your rules.

* To avoid lock you can also forbid the adding of any rules after its first use. If a package has to be sure no other code has created a godb.DB instance it can do it into an `init()` function. Of course the is no guarantee no other code has created a godb.DD instance in an `init()` function... but this has NOT to be done. The only one things needed is a way to return error (or even panic !) if a rules is added to late.

Only one instance mean potential rules conflict... in theory. In practice english is english, plural has to be the same for all. If an edge case exists fallback to `TableName()` is the solution.

To reply to a comment you made *as a DBA* , and about eventual rule conflict : I agree about the use of plural in table names. But I think pluralization by code is more a problem than a solution ;) If somebody want something *rock-solid* he's better to give himself the table name. Any change in plural package or any customization added by another package will not break you mapping. It's my personal point of view (nobody is obliged to agree, of course).

About tests... there are no tests for `tablenamer.go` , and nor tablenamer nor plural made testing about cache.

I'm perhaps punctilious. But I take long time to design / rewrite parts in godb. And I think things are almost done here ;)

Sam.

samonzeweb commented:
Things seems good, just three points : 

* Format the code with `gofmt` 
```
gofmt -w tablenamer/*.go
gofmt -w tablenamer/plural/*.go
```

* `SetDefaultTableNamer` has no comment (rather important as it's the entry point of your feature).
* Drop 1.8 support : update `README.md` and `.travis.yml` , then the CI will not fail.

About the `gofmt` , I use VSCode which has a great support of Go and does the job automatically. Perhaps you can turn on a similar feature in your editor ? It's a so great tool I'm sad there is no equivalent (strong standard and tool) in C#   ;)

derkan commented:
All done!
Also switched to VSCode after previous commit. :+1:

samonzeweb commented:
I just merged your PR, but I found a problem.

The same cache is used to cache diverse things (I saw it due to its re-initialisation I missed before). It's bad as two database could use different naming rules in the same code (probably not the common case, but it could occur then it has to be safe). To be worse it could perhaps lead to race condition if the cache is used and reset by two goroutines at *the same time* (low probability, but > 0) :

```
// sCache is cache for ToSnakeCase function
var sCache = sync.Map{}

[...]

func Snake() NamerFn {
	sCache = sync.Map{}

[...]

func SnakePlural() NamerFn {
	sCache = sync.Map{}
```

I think the simplest way is to :
* use two caches (snake & snake plural).
* initialise them only once at package level.

I can correct it myself, but I just prefers to have your opinion as you write the code.

samonzeweb commented:
Done.
